pytorch_lightning/trainer/trainer.py

```diff
@@ -490,6 +490,38 @@ def _setup_on_init(self, num_sanity_val_steps: int) -> None:
 
         self.num_predict_batches = []
 
+    def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs: Any):
+        r"""
+        Error handling, intended to be used only for main trainer function entry points (fit, validate, test, predict)
+        as all errors should funnel through them
+
+        Args:
+            trainer_fn: one of (fit, validate, test, predict)
+
+            *args/**kwargs: args to be passed to trainer_fn
+        """
+        try:
+            return trainer_fn(*args, **kwargs)
+        except KeyboardInterrupt:
+            rank_zero_warn("Detected KeyboardInterrupt, attempting graceful shutdown...")
+            # user could press Ctrl+c many times... only shutdown once
+            if not self.interrupted:
+                self.state.status = TrainerStatus.INTERRUPTED
+                self.on_keyboard_interrupt()
+                # same treatment as below
+                self.accelerator.teardown()
+        except BaseException:
+            self.state.status = TrainerStatus.INTERRUPTED
+            if distributed_available() and self.world_size > 1:
+                # try syncing remaing processes, kill otherwise
+                self.training_type_plugin.reconciliate_processes(traceback.format_exc())
+            # give accelerators a chance to finish
+            self.accelerator.teardown()
```

Comment:
the difference here is that handling `BaseException` will re-raise the original exception, while the handling for `KeyboardInterrupt` will not. Therefore, I don't think we should teardown the accelerator here. It's likely that doing so could raise an exception on its own, which would mask the original error. 
```suggestion
```